### ISSUE TITLE
Reproducibility Issues

### ISSUE THREAD

**Issue by luke-modern-intelligence** [yongchao97/FRePo#12]
Not sure what I am doing wrong, but I am not able to get near the results reported in the paper when using the google drive checkpoint images in pytorch. Heres what I have done:

- Trained LeNet on mnist_ipc1_llTrue and False
- Adam optimizer with multiple learning rates in the range of .003-.00003. 
- Tried no weight decay or decay of .0001
- Trained for anywhere from 50 to 10,000 epochs

With all these variations, I have only achieved anywhere from 8-20% accuracy on the real mnist testing set. The paper does not mention the need to preprocess the test set. The training images look good to me, match that in the paper (although they have  inverted color w.r.t the paper). 

The labels seem off though. Learned Labels look okay, but when Learned label is false, the labels are the identity matrix minus .1. So it looks like [.9, -.1, -.1, ...] for class 0. I replaced these labels with the identity matrix and was able to get around 70% accuracy on MNIST (very large variance depending on weight initializations). Was hoping to see somewhere around 85-90% accuracy on this but maybe it is just becuase LeNet is a very different architecture than the conv it was distilled on.

I tried to use the exact conv-bn as described in the paper; however, with the checkpoint labels (both true & false) these get 10-20% accuracy. With the identity trick I did for LeNet, I only get around 40% accuracy, not 90+%.

The hyperparameters for these experiments aren't well documented in the paper, so I could be using completely wrong things. Are the any other ideas for what I could be doing wrong for reproducing on MNIST?

**Comment by luke-modern-intelligence** [comment]
Normalization on test data brings ConvBN to 80%, but the label issue is still happening. With my hyperparameters, the identity matrix is far better than any of the labels given.

**Comment by yongchao97** [comment]
Hi Luke,

Thanks for reaching out. It seems there may be some preprocessing data issue in your implementation, but I cannot tell from the description. Have you tried to evaluate using the provided [script]( https://github.com/yongchao97/FRePo/blob/master/script/eval.py)? I have done some sanity checks in the past and found that they worked out fine. Maybe you can contrast your code with mine to see the difference? The label is right in my case since all my networks are trained using MSE loss rather than standard CE loss. 

Generally speaking, I suggest first working with the CONV architecture in the repo, and when you get the expected performance, you can evaluate it on LeNET.

Best,
Yongchao

**Comment by luke-modern-intelligence** [comment]
Thanks Yongchao, I just saw the pytorch branch too, so now I can properly compare code efficiently. As for running your code, I was running into issues with the environment on M1 mac. I needed to use the checkpoints on my own models, so I am trying to get this to work on general implementations. I imagined these distilled images would work pretty much out of the box like standard images with general hyperparameter search, so I see that you be very careful with what you pick. I will have to switch over to your code. 

When you say the "label is right", do you mean your implementation uses a standard identity matrix, or that the identity matrix - .1 is correct for learned_label false? If the latter, why is that the case to have negative values on all other classes?

Also, for larger architecures like ResNet-50/18 for distilling, you do not show the results in the paper but comment that they lead to poor cross-architecture generalization. When evaluating on ResNet, how well do they do as distillers compared to ConvBN? 
ex: ResNet-18 distilled Images -> Evaluate on a fresh ResNet-18 vs. ConvBN distilled Images -> Evaluate on ResNet-18

Thanks for reaching back, Luke
